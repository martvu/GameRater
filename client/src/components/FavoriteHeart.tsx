import { Heart } from 'lucide-react';
import { Button } from './ui/button';
import { useEffect, useState } from 'react';
import { useRecoilState } from 'recoil';
import { userState } from '@/state/atoms';
import { Game } from '@/gql/graphql';
import { useMutation } from '@apollo/client';
import { ADD_FAVORITES, REMOVE_FAVORITES } from '@/lib/mutations';
import { Dialog, DialogContent } from './ui/dialog';
import SignInAlertModal from './SignInAlertModal';

interface FavoriteHeartProps {
  game: Game;
  variant?: 'secondary' | 'ghost' | 'text' | 'outline';
}

/**
 * FavoriteHeart component
 * @param {Game} game - The game object
 * @param {string} variant - The variant of the button (default ghost)
 */
export default function FavoriteHeart({ game, variant }: FavoriteHeartProps) {
  const [addFavorites] = useMutation(ADD_FAVORITES);
  const [removeFavorite] = useMutation(REMOVE_FAVORITES);
  const [user, setUser] = useRecoilState(userState);
  const [isUpdating, setIsUpdating] = useState(false);
  const [isDialogOpen, setIsDialogOpen] = useState(false);

  const toggleFavorite = async (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    event.preventDefault();
    event.stopPropagation();
    // If the user is not signed in, open the sign in alert modal
    if (!user.username) {
      setIsDialogOpen(true);
      return;
    }
    if (isUpdating) return;
    setIsUpdating(true);
    if (!user.username) {
      setIsUpdating(false);
      return;
    }
    const isFavorite = user.favorites?.some(
      favoriteGame => favoriteGame?._id === game._id
    );
    const updatedFavorites = isFavorite
      ? user.favorites?.filter(
          favoriteGame => favoriteGame && favoriteGame._id !== game._id
        )
      : [...(user.favorites || []), game];

    // Optimistically update the UI
    setUser({ ...user, favorites: updatedFavorites });

    try {
      await (isFavorite ? removeFavorite : addFavorites)({
        variables: {
          username: user.username as string,
          gameID: game._id as string,
        },
      });
    } catch (error) {
      console.log('Could not update favorites', error);
      setUser(user);
    } finally {
      setIsUpdating(false);
    }
  };

  const handleSignInSuccess = () => {
    setIsDialogOpen(false);
  };

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  }, [user]);

  const isFavorite = user.favorites?.some(
    favoriteGame => favoriteGame?._id === game._id
  );

  return (
    <>
      <Button
        aria-label="Favorite game button"
        data-testid="favorite-btn"
        onClick={toggleFavorite}
        className="z-10"
        variant={variant || 'ghost'}
        size="round"
      >
        <Heart
          data-testid="heart-icon"
          className={isFavorite ? 'fill-red-600 text-red-600' : 'none'}
        />
      </Button>
      <Dialog open={isDialogOpen} onOpenChange={setIsDialogOpen}>
        <DialogContent>
          <SignInAlertModal
            message={'Please sign in to favorite this game!'}
            onSignInSuccess={handleSignInSuccess}
          />
        </DialogContent>
      </Dialog>
    </>
  );
}
