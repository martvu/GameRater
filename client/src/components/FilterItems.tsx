import { useState } from 'react';
import { Plus, Minus } from 'lucide-react';
import { Checkbox } from './ui/checkbox';
import { Button } from './ui/button';
import { useRecoilState, useSetRecoilState } from 'recoil';
import {
  pageState,
  selectedGenresState,
  selectedPlatformsState,
} from '@/state/atoms.ts';
import { Genre, Platform } from '@/gql/graphql';
import { Label } from './ui/label';

interface FilterItemsProps {
  filters: Platform[] | Genre[];
  filterType: 'platforms' | 'genres';
}

export default function FilterItems({ filters, filterType }: FilterItemsProps) {
  const [numItemsToShow, setNumItemsToShow] = useState(10);
  const [selectedPlatforms, setSelectedPlatforms] = useRecoilState(
    selectedPlatformsState
  );
  const [selectedGenres, setSelectedGenres] =
    useRecoilState(selectedGenresState);
  const setCurrentPage = useSetRecoilState(pageState);
  const showMore = () => {
    setNumItemsToShow(numItemsToShow + 10);
  };

  const handleCheckboxChange = (filterId: number, isChecked: boolean) => {
    if (filterType === 'platforms') {
      setSelectedPlatforms(
        isChecked
          ? [...selectedPlatforms, filterId]
          : selectedPlatforms.filter(id => id !== filterId)
      );
    } else {
      setSelectedGenres(
        isChecked
          ? [...selectedGenres, filterId]
          : selectedGenres.filter(id => id !== filterId)
      );
    }
    setCurrentPage(1); // Reset page to 1 when a filter is selected
  };

  const showLess = () => setNumItemsToShow(10);
  return (
    <div>
      <h2 className="my-2 text-left font-semibold capitalize text-foreground">
        {filterType}
      </h2>
      <div className="ml-1 flex flex-col">
        {filters?.slice(0, numItemsToShow).map(item => (
          <div className="my-1 flex items-center space-x-2" key={item?.name}>
            <Checkbox
              id={`filter-item-${item?.name}`}
              onCheckedChange={checked => {
                if (typeof item?.id === 'number')
                  handleCheckboxChange(item.id, checked.valueOf() as boolean);
              }}
              checked={
                filterType === 'platforms'
                  ? selectedPlatforms.includes(item?.id ?? 0)
                  : selectedGenres.includes(item?.id ?? 0)
              }
            />
            <Label
              htmlFor={`filter-item-${item?.name}`}
              className={`text-left font-normal tracking-tight ${
                (filterType === 'platforms' &&
                  selectedPlatforms.includes(item?.id ?? 0)) ||
                (filterType === 'genres' &&
                  selectedGenres.includes(item?.id ?? 0))
                  ? 'text-foreground'
                  : 'text-muted-foreground'
              }`}
            >
              {item?.name}
              {/* {item?.gamesCount && ` (${item?.gamesCount})`} */}
            </Label>
          </div>
        ))}
      </div>
      {filters && numItemsToShow < filters.length ? (
        <Button
          variant={'text'}
          onClick={showMore}
          className="flex w-36 items-center pl-0 text-green-400"
        >
          <Plus className="inline-block h-6" />
          <span>Show More</span>
        </Button>
      ) : numItemsToShow > 10 && filters.length > 10 ? (
        <Button
          variant={'text'}
          onClick={showLess}
          className="flex w-36 items-center pl-0 text-red-400"
        >
          <Minus className="inline-block h-6" />
          <span>Show Less</span>
        </Button>
      ) : (
        <div></div>
      )}
    </div>
  );
}
