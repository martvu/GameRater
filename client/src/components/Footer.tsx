export default function Footer() {
  return (
    <footer className="mt-5 h-16 w-full border-t text-muted-foreground">
      <div className="container mx-auto px-6 pt-5 text-left text-sm">
        Copyright 2023 GameRater
      </div>
    </footer>
  );
}
