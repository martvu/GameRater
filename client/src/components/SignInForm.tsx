import { zodResolver } from '@hookform/resolvers/zod';
import * as z from 'zod';
import { Button } from '@/components/ui/button';

import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { useForm } from 'react-hook-form';
import { useRecoilState } from 'recoil';
import { userState } from '@/state/atoms';
import { gql, useMutation } from '@apollo/client';
import { useToast } from '@/components/ui/use-toast.ts';

const SIGN_IN_OR_CREATE_USER = gql`
  mutation SignInOrCreateUser($userInput: UserInput) {
    signInOrCreateUser(userInput: $userInput) {
      _id
      username
      favorites {
        _id
      }
      reviews {
        _id
      }
    }
  }
`;

const formSchema = z.object({
  username: z
    .string()
    .min(2, {
      message: 'Username must be at least 2 characters.',
    })
    .max(15, {
      message: 'Username must be at most 15 characters.',
    }),
});

interface SignInFormProps {
  onClose: () => void;
}

export function SignInForm({ onClose }: SignInFormProps) {
  const [signInOrCreateUser] = useMutation(SIGN_IN_OR_CREATE_USER);
  const setUser = useRecoilState(userState)[1];

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      username: '',
    },
  });
  const { toast } = useToast();

  async function onSubmit(values: z.infer<typeof formSchema>) {
    try {
      onClose();
      const { data } = await signInOrCreateUser({
        variables: {
          userInput: {
            username: values.username.trim(),
          },
        },
      });
      setUser(data.signInOrCreateUser);
      localStorage.setItem('user', JSON.stringify(data.signInOrCreateUser));
      toast({
        title: 'Signed in successfully',
        description: `Welcome, ${data.signInOrCreateUser.username}!`,
      });
    } catch (error) {
      console.log('Could not create user', error);
      toast({
        variant: 'destructive',
        title: 'Could not sign in',
        description: 'Please try again.',
      });
    }
    form.reset();
  }

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-2">
        <FormDescription className="w-60">
          We just need your username, and you're good to go!
        </FormDescription>
        <FormField
          control={form.control}
          name="username"
          render={({ field }) => (
            <FormItem>
              <FormLabel>Username</FormLabel>
              <div className="flex items-center gap-4">
                <FormControl>
                  <Input
                    data-testid="username-input"
                    placeholder="Your username"
                    {...field}
                    className="h-10 max-w-[200px]"
                    maxLength={15}
                  />
                </FormControl>
                <Button data-testid="form-sign-in-button" type="submit">Sign In</Button>
              </div>
              <FormMessage />
            </FormItem>
          )}
        />
      </form>
    </Form>
  );
}
