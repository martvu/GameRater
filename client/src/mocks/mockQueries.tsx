import {
  ADD_FAVORITES,
  CREATE_REVIEW,
  REMOVE_FAVORITES,
  SIGN_IN_OR_CREATE_USER,
} from '@/lib/mutations';
import {
  GET_GAMES,
  GET_GAME,
  GET_FILTERS,
  GET_GAME_PLATFORMS,
  GET_SEARCH_SUGGESTIONS,
} from '@/lib/queries';

const getGamesMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: undefined,
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesMockedKeyword = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            user_rating: 4.5,
          },
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getGamesSignedInMock = {
  request: {
    query: GET_GAMES,
    variables: {
      limit: 24,
      offset: 0,
      sortBy: { field: 'first_release_date', order: 'desc' },
      platforms: [],
      genres: [],
      query: 'mockedKeyword',
      userId: '1',
      showFavorites: false,
      showReviewedGames: false,
    },
  },
  result: {
    data: {
      search: {
        count: 1,
        games: [
          {
            _id: '1',
            aggregated_rating: 85,
            first_release_date: 1620000000,
            summary: 'Example game summary',
            cover_image_id: 'cover1',
            name: 'Example Game',
            platforms: [1, 2],
            genres: [3, 4],
          },
          // Add more game objects as needed
        ],
        filters: {
          genres: [1, 2], // Example genre IDs
          platforms: [3, 4], // Example platform IDs
        },
      },
    },
  },
};

const getFiltersMock = {
  request: {
    query: GET_FILTERS,
    variables: {},
  },
  result: {
    data: {
      getGenres: [
        { id: 1, name: 'Genre 1' },
        { id: 2, name: 'Genre 2' },
      ],
      getPlatforms: [
        { id: 1, name: 'Platform 1' },
        { id: 3, name: 'Platform 3' },
        { id: 4, name: 'PC (Microsoft Windows)' },
        { id: 2, name: 'Nintendo Switch' },
        { id: 5, name: 'PlayStation 5' },
        { id: 6, name: 'Xbox Series X|S' },
      ],
    },
  },
};

const getGamePlatformsMock = {
  request: {
    query: GET_GAME_PLATFORMS,
    variables: { id: '1' },
  },
  result: {
    data: {
      getGame: {
        name: 'Example Game',
        platforms: [
          {
            id: 3,
            name: 'Platform 3',
          },
          {
            id: 4,
            name: 'PC (Microsoft Windows)',
          },
        ],
      },
    },
  },
};

const signInMock = {
  request: {
    query: SIGN_IN_OR_CREATE_USER,
    variables: {
      userInput: {
        username: 'testuser',
      },
    },
  },
  result: {
    data: {
      signInOrCreateUser: {
        _id: '1',
        username: 'testuser',
        favorites: [],
        reviews: [],
      },
    },
  },
};

const getGameMock = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: '',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};

const getGameMock2 = {
  request: {
    query: GET_GAME,
    variables: {
      id: '1',
      limit: 5,
      offset: 0,
      username: 'testUser',
    },
  },
  result: {
    data: {
      getGame: {
        _id: '1',
        name: 'Example Game',
        summary: 'This is an example game summary.',
        imageId: 'image1',
        first_release_date: 1620000000,
        aggregatedRating: 85,
        user_rating: 4.5,
        platforms: [{ name: 'PC' }],
        genres: [{ name: 'Adventure' }],
        reviews: {
          count: 2,
          reviews: [
            {
              _id: 'review1',
              user: 'user1',
              title: 'Great Game',
              content: 'This is a great game because...',
              rating: 5,
              platform: 'PC',
              gameID: '1',
            },
            {
              _id: 'review2',
              user: 'user2',
              title: 'Good Game',
              content: 'This is a good game because...',
              rating: 4,
              platform: 'PC',
              gameID: '1',
            },
          ],
          userHasReviewed: false,
        },
      },
    },
  },
};
const addFavoritesMock = {
  request: {
    query: ADD_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      addFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [{ _id: '1' }],
      },
    },
  },
};

const removeFavoritesMock = {
  request: {
    query: REMOVE_FAVORITES,
    variables: { username: 'testUser', gameID: '1' },
  },
  result: {
    data: {
      removeFavorites: {
        _id: '1',
        username: 'testUser',
        favorites: [],
      },
    },
  },
};

const createReviewMock = {
  request: {
    query: CREATE_REVIEW,
    variables: {
      reviewInput: {
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        user: 'testUser',
        gameID: '1',
      },
    },
  },
  result: {
    data: {
      createReview: {
        _id: 'review1',
        user: 'testUser',
        title: 'Great Game',
        content: 'This is a great game because...',
        rating: 5,
        platform: 'PC (Microsoft Windows)',
        gameID: '1',
      },
    },
  },
};

const searchbarSuggestionsMock = {
  request: {
    query: GET_SEARCH_SUGGESTIONS,
    variables: {
      query: 'example',
    },
  },
  result: {
    data: {
      getSearchSuggestions: [
        {
          _id: '1',
          name: 'Example Game',
          cover_image_id: 'cover1',
        },
        {
          _id: '2',
          name: 'Example Game 2',
          cover_image_id: 'cover2',
        },
      ],
    },
  },
};

const allMocks = [
  createReviewMock,
  getGamePlatformsMock,
  getGamesMockedKeyword,
  addFavoritesMock,
  removeFavoritesMock,
  getGamesMock,
  getFiltersMock,
  signInMock,
  getGamesSignedInMock,
  getGameMock,
  getGameMock2,
  searchbarSuggestionsMock,
  // Add more mocks as needed for other queries
];

export {
  allMocks,
  getGamesMock,
  getGamesSignedInMock,
  getFiltersMock,
  signInMock,
};
