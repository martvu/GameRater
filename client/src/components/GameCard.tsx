import { Star } from 'lucide-react';
import imageNotFound from '@/assets/img-fallback.svg';

import {
  Card,
  CardContent,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Link } from 'react-router-dom';
import { Game } from '@/gql/graphql';
import { useQuery } from '@apollo/client';
import { gql } from '../gql/';
import Metascore from './Metascore';
import FavoriteHeart from './FavoriteHeart';
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from './ui/tooltip';

interface GameCardProps {
  game: Game;
}

const GET_AVG_RATING = gql(`
  query GetAvgRating($gameID: ID!) {
    getAvgRating(gameID: $gameID)
  }
`);

export function GameCard({ game }: GameCardProps) {
  const {
    _id: id,
    name,
    aggregated_rating: aggregatedRating,
    cover_image_id: imageId,
  } = game;
  const { data } = useQuery(GET_AVG_RATING, {
    variables: { gameID: id as string },
  });

  const handleImageError = (e) => {
    e.target.src = imageNotFound; // Set the source to your fallback image
  };
  const rating = data?.getAvgRating;
  const coverImageUrl = imageId
    ? `https://images.igdb.com/igdb/image/upload/t_cover_big/${imageId}.jpg`
    : ('' as string);

  return (
    <Card className="relative h-[310px] min-w-[175px] max-w-[260px] p-0 duration-300 overflow-hidden hover:scale-105 sm:h-[410px] sm:min-w-[260px]">
      <Link to={`/game/${id}`} aria-label={`Link to ${name} detail page`}>
        <CardHeader className="mb-2 h-[240px] w-[175px] overflow-hidden sm:h-[320px] sm:w-[260px]">
          <div className="flex flex-1 items-center justify-center">
            {imageId && (
              <img
                src={coverImageUrl}
                alt={name as string}
                className="object-cover fill-green-300"
                onError={handleImageError}
              />
            )}
          </div>
        </CardHeader>
        <CardContent className="relative px-3 text-start">
          <TooltipProvider>
            <Tooltip>
              <TooltipTrigger asChild>
                <CardTitle className="my-1 line-clamp-1 max-w-[150px] py-1 text-sm duration-300 hover:opacity-50 sm:max-w-[240px] sm:text-lg ">
                  {name}
                </CardTitle>
              </TooltipTrigger>
              <TooltipContent>
                <p className="text-xs md:text-sm">{name}</p>
              </TooltipContent>
            </Tooltip>
          </TooltipProvider>
        </CardContent>
      </Link>
      <CardFooter className="absolute bottom-0 left-0 mt-auto h-[40px] w-full px-3 pb-2">
        <div className="relative flex h-full w-full items-center">
          <div className="absolute left-0 flex items-center">
            <Star className="mr-1 h-5 fill-yellow-400 text-yellow-400" />
            <p>{rating === 0 ? '-' : rating}</p>
          </div>
          <div className="absolute right-10 flex items-center gap-2">
            <Metascore
              metascore={aggregatedRating ? aggregatedRating : undefined}
            />
          </div>
          <div className="absolute right-0">
            <FavoriteHeart game={game} />
          </div>
        </div>
      </CardFooter>
    </Card>
  );
}
