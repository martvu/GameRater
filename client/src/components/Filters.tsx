import { useState } from 'react';
import { ChevronLeft, ListFilter } from 'lucide-react';
import FilterItems from './FilterItems';
import { Button } from './ui/button';
import { cn } from '@/lib/utils';
import { GET_FILTERS } from '@/lib/queries';
import { useQuery } from '@apollo/client';
import { Genre, Platform } from '@/gql/graphql';
import { useRecoilValue } from 'recoil';
import {
  genresListState,
  platformsListState,
  selectedGenresState,
  selectedPlatformsState,
} from '@/state/atoms';
import { ScrollArea } from './ui/scroll-area';

/**
 * Filters component
 * Displayed on large screens (non-mobile)
 */
export default function Filters() {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const genreList = useRecoilValue(genresListState);
  const platformList = useRecoilValue(platformsListState);
  const selectedGenres = useRecoilValue(selectedGenresState);
  const selectedPlatforms = useRecoilValue(selectedPlatformsState);

  const toggleCollapse = () => {
    setIsCollapsed(!isCollapsed);
  };
  const { data } = useQuery(GET_FILTERS);

  const preferredOrder = [
    'Nintendo Switch',
    'PC (Microsoft Windows)',
    'PlayStation 5',
    'Xbox Series X|S',
  ];

  // Custom sorting function to sort platforms by preferred order
  const sortByPreference = (a: Platform, b: Platform) => {
    const aIndex = preferredOrder.indexOf(a.name as string);
    const bIndex = preferredOrder.indexOf(b.name as string);

    if (aIndex !== -1 && bIndex !== -1) {
      return aIndex - bIndex; // Both are preferred, sort by preferred order
    } else if (aIndex !== -1) {
      return -1; // Only a is preferred
    } else if (bIndex !== -1) {
      return 1; // Only b is preferred
    }
    return 0; // Neither is preferred, maintain original order
  };

  // Sort the platforms
  // Create a copy of the platforms array before sorting
  const platformsCopy: Platform[] = [...(data?.getPlatforms ?? [])];
  const genresCopy: Genre[] = [...(data?.getGenres ?? [])];
  // Now sort the copied array
  const sortedPlatforms: Platform[] = platformsCopy.sort(sortByPreference);

  // Filter only platforms in platformList (when search limits platforms) or all platforms if platformList is empty
  const filteredPlatforms = sortedPlatforms.filter(platform => {
    if (platformList.length === 0) return true;
    return (
      platformList.includes(platform.id as number) ||
      selectedPlatforms.includes(platform.id as number)
    );
  });

  // Filter only genres in genreList or all genres if genreList is empty
  const filteredGenres = genresCopy.filter(genre => {
    if (genreList.length === 0) return true;
    return (
      genreList.includes(genre.id as number) ||
      selectedGenres.includes(genre.id as number)
    );
  });

  // Apply transition to the filter menu
  const filterContainerClasses = cn(
    'flex h-full flex-col items-start overflow-hidden text-left transition-all duration-300 ease-in-out',
    isCollapsed ? 'w-0' : 'w-[240px]' // Toggle between full width and collapsed state
  );

  return (
    <>
      <section className={filterContainerClasses}>
        <ScrollArea className="h-full pb-4 pr-2">
          <section className="pb-4 pl-5 md:pb-32">
            <h1
              className={`mt-4 text-left text-xl font-bold tracking-wider text-foreground ${
                isCollapsed ? 'hidden' : ''
              }`}
            >
              Filter By
            </h1>
            {!isCollapsed && data && (
              <>
                <FilterItems
                  filters={filteredPlatforms}
                  filterType="platforms"
                />
                <FilterItems filters={filteredGenres} filterType="genres" />
              </>
            )}
          </section>
        </ScrollArea>
      </section>
      <Button
        className="fixed bottom-8 left-8 z-50 m-1 hidden md:flex"
        variant="secondary"
        onClick={toggleCollapse}
        aria-label={isCollapsed ? 'show filters' : 'hide filters'}
      >
        {isCollapsed ? <ListFilter /> : <ChevronLeft size={18} />}
        {!isCollapsed && <span className="ml-2">Hide filters</span>}
      </Button>
    </>
  );
}
