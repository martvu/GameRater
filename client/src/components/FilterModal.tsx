import { Button } from '@/components/ui/button';
import {
  Dialog,
  DialogClose,
  DialogContent,
  DialogFooter,
  DialogTrigger,
} from '@/components/ui/dialog';
import Filters from './Filters';

/**
 * FilterModal component
 * Displayed on small screens (mobile devices)
 */
export default function FilterModal() {
  return (
    <Dialog>
      <DialogTrigger asChild>
        <Button variant="outline" className="w-[150px]">
          Filters
        </Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-[425px]">
        <Filters />
        <DialogFooter>
          <DialogClose asChild>
            <Button type="submit">Apply Filter</Button>
          </DialogClose>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
