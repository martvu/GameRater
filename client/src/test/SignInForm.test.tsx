import { SignInForm } from "@/components/SignInForm";
import { allMocks } from "@/mocks/mockQueries";
import { MockedProvider } from "@apollo/client/testing";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { RecoilRoot } from "recoil";
import { vi } from "vitest";


const onCloseMock = vi.fn();

const renderSignInForm = () => {
  return (
    render(
    <MockedProvider mocks={allMocks} addTypename={false}>
      <RecoilRoot>
        <SignInForm onClose={onCloseMock} />
      </RecoilRoot>
    </MockedProvider>
    )
  );
};
describe('SignInForm Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderSignInForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    renderSignInForm();
    expect(document.querySelector('form')).toBeInTheDocument();
    expect(document.querySelector('input[name="username"]')).toBeInTheDocument();
    expect(document.querySelector('button[type="submit"]')).toBeInTheDocument();
  });

  it('signs in', async () => {
    renderSignInForm();
    await userEvent.type(screen.getByLabelText('Username'), 'testuser{enter}');
    expect(onCloseMock).toHaveBeenCalled();
  });
});