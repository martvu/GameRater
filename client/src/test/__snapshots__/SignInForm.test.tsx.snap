// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignInForm Component > matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="space-y-2"
  >
    <p
      class="text-sm text-muted-foreground w-60"
      id="undefined-form-item-description"
    >
      We just need your username, and you're good to go!
    </p>
    <div
      class="space-y-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Username
      </label>
      <div
        class="flex items-center gap-4"
      >
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-10 max-w-[200px]"
          data-testid="username-input"
          id=":r0:-form-item"
          maxlength="15"
          name="username"
          placeholder="Your username"
          value=""
        />
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 active:bg-primary/70 h-10 px-4 py-2"
          data-testid="form-sign-in-button"
          type="submit"
        >
          Sign In
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
